app/anti_virus_check.py

```diff
@@ -0,0 +1,136 @@
+            "Unable to get results of Anti-virus scan for case id {} and file {} ".format(payload.case_id,
+                                                                                          payload.file_name))
+        raise RetryableError()
+
+    def _add_api_key(self, headers):
+        # the API key is only needed for the online cloud service
+        if settings.ANTI_VIRUS_API_KEY:
+            self.bound_logger.debug("Setting A/V API key")
+            headers['apikey'] = settings.ANTI_VIRUS_API_KEY
+
+    def _check_av_response(self, response):
+        if response.status_code != 200:
+            if response.status_code == 401:
+                self.bound_logger.critical("Invalid OPSWAT API Key - unable to continue")
+                raise RetryableError()
+            elif response.status_code == 404:
+                # this could mean that the primary A/V server has failed and we've failed over to the backup
+                # in this scenario we need to start over
+                self.bound_logger.critical("OPSWAT AV does not know about this scan - the primary server may have failed")
+                raise RetryableError()
+            elif 400 <= response.status_code < 500:
+                raise BadMessageError()
+            elif 300 < response.status_code >= 500:
+                raise RetryableError()
+        else:
+            self.bound_logger.info("Response from AV OK")
+
+    def _send_for_anti_virus_check(self, filename, contents):
+        url = settings.ANTI_VIRUS_BASE_URL + "file"
+        headers = {
+            "filename": filename,
+        }
+        self._add_api_key(headers)
+
+        self.bound_logger.info("Sending for A/V scan", url=url)
+        response = requests.post(url=url, headers=headers, data=contents)
+
+        self._check_av_response(response)
+
+        self.bound_logger.info("Response received {}".format(response.text))
+        result = response.json()
+
+        if result.get("err"):
+            self.bound_logger.error("Unable to send file for anti virus scan: {}".format(result.get("err")))
+            self.bound_logger.info("Waiting before attempting again")
+            time.sleep(settings.ANTI_VIRUS_WAIT_TIME)
+            self.bound_logger.info("Return message to rabbit")
+            raise RetryableError()
+        else:
+            data_id = result.get("data_id")
+            self.bound_logger.info("File sent successfully for anti virus scan", data_id=data_id)
+            return data_id
+
+    def _get_anti_virus_result(self, data_id):
+        url = settings.ANTI_VIRUS_BASE_URL + "file/" + data_id
+        headers = {}
+        self._add_api_key(headers)
+
+        self.bound_logger.info("Getting result for A/V scan", url=url)
+
+        response = requests.get(url=url, headers=headers)
+
+        self._check_av_response(response)
+
+        self.bound_logger.info("Response from AV {}".format(response.text))
+
+        result = response.json()
```

Comment:
Are the results guaranteed to be decodable? If not, watch out for JSONDecode exceptions.

Reply:
Done